src/state.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm state module."""
+
+import itertools
+import logging
+from abc import ABC, abstractmethod
+from typing import Optional
+
+import ops
+from pydantic import AnyUrl, BaseModel, ValidationError
+
+from config import CharmConfig
+
+logger = logging.getLogger(__name__)
+
+
+class RecoverableStateError(Exception):
```

Comment:
I would rename this. To me, it look like this exception is used for incorrect configuration